### PR TITLE
Feature: paginate events -  PCD-387

### PR DESCRIPTION
**Pull request by sebasfavaron**
## Summary

Instead of loading ALL events on first page load and then filtering, gallery now fetches paginated results from the backend (according to the search value and order filters present)
This fixes:
- slow first load due to big network usage
- events not being found on search
- filtering consistency

The last two issues came from a bug where not all events were actually being loaded, so searching some events failed and filtering did not yield a list consistent with the backend data

Note: This is NOT the full backend solution as data merging from the api and subgraphs is still ocurring in the frontend, but it should aliviate some big problems on gallery
Note2: Surrounded anchor tags with object tags as a hack to silence '\<a\> cannot appear as a descendant of \<a\>' warnings

### Average load time until events are loaded (15 runs)
Before: 15,62s
After: 3,45s

### Searches:
Before:
image
After:
image

### Incorrect filtering:
Before (thinks 9999 is the biggest ID)
image
After (correctly orders IDs)
image

### Missing events:
Before (Mutant Apes and other events missing)
image
After
image

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

## Testing

- Compare loading times with poap.gallery and verify that they are faster
- Search for an event that didn't appear on poap.gallery and verify it appears now (ex: 'chainlink ico')

## Ticket

https://poap-devs.atlassian.net/jira/software/projects/PCD/boards/9?assignee=5c4bc8d4043b4f5d172a6fdd&selectedIssue=PCD-387
https://poap-devs.atlassian.net/jira/software/projects/PCD/boards/9?assignee=5c4bc8d4043b4f5d172a6fdd&selectedIssue=PCD-475

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I didn't commit unnecessary logs

### CODE REVIEW

File: src/components/activityTable.js [hunk 1]
```diff
@@ -107,18 +107,18 @@ function Transfer({transfer}) {
...
-                <a href={"https://app.poap.xyz/scan/" + transfer.from.id} target="_blank"  rel="noopener noreferrer"> {transfer.from.id.substring(0, 8) + '…'} </a> to 
-                <a href={"https://app.poap.xyz/scan/" + transfer.to.id} target="_blank"  rel="noopener noreferrer"> {transfer.to.id.substring(0, 8) + '…'} </a>
+                <object><a href={"https://app.poap.xyz/scan/" + transfer.from.id} target="_blank"  rel="noopener noreferrer"> {transfer.from.id.substring(0, 8) + '…'} </a></object> to
+                <object><a href={"https://app.poap.xyz/scan/" + transfer.to.id} target="_blank"  rel="noopener noreferrer"> {transfer.to.id.substring(0, 8) + '…'} </a></object>
                  on {transfer.network}
               </span> :
               (type === 'Migration')?
               <span> POAP migrated to 
-                <a href={"https://app.poap.xyz/scan/" + transfer.to.id} target="_blank"  rel="noopener noreferrer"> {transfer.to.id.substring(0, 16) + '…'} </a>
+                <object><a href={"https://app.poap.xyz/scan/" + transfer.to.id} target="_blank"  rel="noopener noreferrer"> {transfer.to.id.substring(0, 16) + '…'} </a></object>
```
Comment (pablojp) [inline src/components/activityTable.js:121]: the URL can be a constant. 
Why is an object tag used?

File: src/components/dropdown.js [hunk 2]
```diff
@@ -1,23 +1,39 @@
...
+  useEffect(() => {
+    if(disable) {
+      setDisplayMenu(false)
+      reset()
+    }
+  }, [disable]) /* eslint-disable-line react-hooks/exhaustive-deps */
+
   let optionsElements = []
   for(let i = 0; i < options.length; i++) {
+    const option = options[i]
```
Comment (pablojp) [inline src/components/dropdown.js:31]: What does the option variable contain? Is the name descriptive?

File: src/pages/activity.js [hunk 3]
```diff
@@ -85,28 +84,26 @@ export default function Activity() {
...
+          })
+          // Sort from newer to older
+          .sort((a, b) => {
+            return b.timestamp - a.timestamp
+          })
+          // Only show a few
+          .slice(0, transferLimit)
+      setTransfers(_transfers)
+    }
+    setNewTransfers().then()
```
Comment (pablojp) [inline src/pages/activity.js:105]: Exception failure needs to be handled

File: src/components/dropdown.js [hunk 4]
```diff
@@ -1,23 +1,39 @@
-import React, { useState } from 'react';
+import React, {useEffect, useState} from 'react';
 import Arrow from '../assets/images/angle_down.svg';
 
-export default function Dropdown({title, options, defaultOption, onClickOption}) {
+export default function Dropdown({options, disable, defaultOption, onClickOption}) {
   
-  const [_title, setTitle] = useState(title)
+  const [_title, setTitle] = useState(defaultOption.name)
```
Comment (maurogava) [inline src/components/dropdown.js:6]: no to change now but I wouldn't use the underscore for naming a variable.

File: src/components/dropdown.js [hunk 5]
```diff
@@ -1,23 +1,39 @@
...
   const [displayMenu, setDisplayMenu] = useState(false)
 
   const toggleDisplayMenu = () => {
     setDisplayMenu(!displayMenu)
```
Comment (maurogava) [inline src/components/dropdown.js:10]: I know you didn't write this but for preventing race conditions associated with closures the code should be written as follows:
```
const toggleDisplayMenu = () => setDisplayMenu((prevState) => !prevState)
```

File: src/components/dropdown.js [hunk 6]
```diff
@@ -1,23 +1,39 @@
...
+
+  useEffect(() => {
+    if(disable) {
+      setDisplayMenu(false)
+      reset()
+    }
+  }, [disable]) /* eslint-disable-line react-hooks/exhaustive-deps */
+
   let optionsElements = []
   for(let i = 0; i < options.length; i++) {
```
Comment (maurogava) [inline src/components/dropdown.js:30]: Why don't we use a map?

File: src/components/dropdown.js [hunk 7]
```diff
@@ -1,23 +1,39 @@
...
+    const option = options[i]
     optionsElements.push(
-      <div className={`option ${options[i] === defaultOption?'active':''}`} onClick={() => {
-        setTitle(options[i])
+      <div key={option.name} className={`option`} onClick={() => {
```
Comment (maurogava) [inline src/components/dropdown.js:33]: Why are you using a template literal in the `className`?

File: src/pages/activity.js [hunk 8]
```diff
@@ -85,28 +84,26 @@ export default function Activity() {
...
+  }, [daitransfers, mainnetTransfers])
```
Comment (maurogava) [inline src/pages/activity.js:106]: Not sure about business logic but if `daitransfers` and `mainnetTransfers ` never change, for example if the `useEffects` where they are set fails it won't trigger this `useEffetc`. Plus if both retrieve information correctly this code will be run twice.

File: src/pages/gallery.js [hunk 9]
```diff
@@ -17,134 +17,101 @@ import FailedSearch from '../assets/images/failed-search.svg'
...
+  const [moreToLoad, setMoreToLoad] = useState(true)
 
-  const orderTypeOptions = ['Date', 'Id', 'City', 'Holders', 'Transfers']
-  const [orderType, setOrderType] = useState(orderTypeOptions[0]);
-  const orderDirectionOptions = ['High to Low', 'Low to High']
-  const [orderDirection, setOrderDirection] = useState(orderDirectionOptions[0]);
+  const initialOrderType = OrderType.date
+  const initialOrderDirection = OrderDirection.descending
+  const [orderType, setOrderType] = useState(initialOrderType)
+  const [orderDirection, setOrderDirection] = useState(initialOrderDirection)
```
Comment (maurogava) [inline src/pages/gallery.js:46]: @pablojp This Component needs to be refactored at some point. It's doing too many things and handling a lot of logic. We are also using too many `useStates` and this component is probable re-rendered a lot of times.

File: src/pages/gallery.js [hunk 10]
```diff
@@ -17,134 +17,101 @@ import FailedSearch from '../assets/images/failed-search.svg'
...
   useEffect(() => {
+    if (searchStatus === SEARCH_STATUS.Searching) {
+      setSearchStatus(events.length ? SEARCH_STATUS.Success : SEARCH_STATUS.Failed)
+    }
+
+    if (page > 0 && events.length === items.length) {
+      setMoreToLoad(false)
+    } else {
+      setMoreToLoad(true)
+    }
```
Comment (maurogava) [inline src/pages/gallery.js:82]: you can write this as follows:

```
setMoreToLoad(!(!page && events.length === items.length))
// or applying Morgan's Law
setMoreToLoad(page || events.length !== items.length)
```

File: src/pages/gallery.js [hunk 11]
```diff
@@ -17,134 +17,101 @@ import FailedSearch from '../assets/images/failed-search.svg'
...
     setSearchValue(value);
-    debounceHandleSearch(value, items)
-  };
-  const handleNewSearchValue = (value, items) => {
-    if (value && value.length > 2) {
-      const filteredItems = items.filter((item) => {
-        return item.name.toLowerCase().indexOf(value.toLowerCase()) !== -1;
-      });
-      setSearch(filteredItems);
+    if (value?.length <= 2) {
```
Comment (maurogava) [inline src/pages/gallery.js:97]: This `2` value should be a constant

File: src/pages/gallery.js [hunk 12]
```diff
@@ -17,134 +17,101 @@ import FailedSearch from '../assets/images/failed-search.svg'
...
+      setSearchStatus(SEARCH_STATUS.NoSearch)
     } else {
-      setSearch(undefined);
+      setSearchStatus(SEARCH_STATUS.Searching)
+    }
+  }
+  const handleNewSearchValue = (value) => {
+    if (value && value.length > 2) {
```
Comment (maurogava) [inline src/pages/gallery.js:104]: Same as above

File: src/store/api.js [hunk 13]
```diff
@@ -2,9 +2,64 @@ export const XDAI_SUBGRAPH_URL = process.env.REACT_APP_XDAI_SUBGRAPH_URL;
...
+  }
+  if (offset !== undefined && offset >= 0) {
+    url += `&offset=${offset}`
+  }
+  if (orderBy?.type && orderBy?.order) {
+    url += `&sort_field=${orderBy.type}&sort_dir=${orderBy.order}`
+  }
+  if (privateEvents !== undefined) {
+    url += `&private_event=${privateEvents}`
+  }
```
Comment (maurogava) [inline src/store/api.js:61]: For security reasons we should use the URL API (https://developer.mozilla.org/en-US/docs/Web/API/URL) and build the whole URL with that including the query params

File: src/store/api.js [hunk 14]
```diff
@@ -216,4 +303,9 @@ export async function validateMigrations(migrations) {
     })
   })
   return res2.json()
+}
+
+export async function getTop3Events() {
+  const res = await fetch(`${POAP_API_URL}/top-3-events`)
+  return res.json()
 }
```
Comment (maurogava) [inline src/store/api.js:311]: missing new line

File: src/store/index.js [hunk 15]
```diff
@@ -38,14 +47,37 @@ const eventsSlice = createSlice({
...
+      if (page === 0) {
+        state.currentInvalidResults = invalid
+        state.events = poapEvents
+      } else {
+        state.currentInvalidResults += invalid
+        poapEvents.forEach(poapE => {
+          const match = state.events.find(e => e.id === poapE.id)
+          if (match) {
+            match.tokenCount += poapE.tokenCount
+            match.transferCount += poapE.transferCount
```
Comment (maurogava) [inline src/store/index.js:69]: Aren't we doing a mutation here?

**Comment by maurogava** [comment]
Good job @sebasfavaron, reducing the average loading time by 5x is great!

File: src/pages/activity.js [hunk 16]
```diff
@@ -85,28 +84,26 @@ export default function Activity() {
   }, []);
```
Comment (maurogava) [inline src/pages/activity.js:84]: @pablojp The logic of the above two useEffects that is almost the same shouldn't be here, it should in the API logic probably using a `promiseAll` to retrieve both data and the parse it. And here only set the state referring to the UI.

File: src/pages/activity.js [hunk 17]
```diff
@@ -85,28 +84,26 @@ export default function Activity() {
...
-      })
-      // Sort from newer to older
-      .sort((a, b) => {
-        return b.timestamp - a.timestamp
-      })
-      // Only show a few
-      .slice(0, transferLimit)
-    setTransfers(_transfers)
-  }, [daitransfers, mainnetTransfers, privateEvents])
+    const setNewTransfers = async () => {
```
Comment (maurogava) [inline src/pages/activity.js:87]: @pablojp Thinking of a future refactor this logic probably shouldn't be here. Probably it should be within the logic refactor that I mentioned above.

**Comment by maurogava** [comment]
@pablojp we should discuss about the use of `// eslint-disable-next-line react-hooks/exhaustive-deps`. It is not a good practice and we seem to be using it everywhere.

File: src/components/activityTable.js (hunk 1 above)
Comment (sebasfavaron) [reply src/components/activityTable.js:121]: added note in PR description

File: src/store/index.js (hunk 15 above)
Comment (sebasfavaron) [reply src/store/index.js:69]: If event is already loaded instead of adding it we aggregate the data

**Review by maurogava — APPROVED** [review]

**Review by pablojp — APPROVED** [review]

**Pull request merged**